src/envs/screws.py

```diff
@@ -0,0 +1,509 @@
+
+    def _surface_xy_is_valid(self, x: float, y: float,
+                             existing_xys: Set[Tuple[float, float]]) -> bool:
+        # First, check that there isn't another screw at this location.
+        for (existing_x, existing_y) in existing_xys:
+            if (abs(existing_x - x) < self._screw_width
+                    and abs(existing_y - y) < self._screw_height):
+                return False
+        # Next, check that the location is valid.
+        y_is_valid = y == self.rz_y_lb + self._screw_height / 2.0
+        x_is_valid = (self.rz_x_lb + self._screw_width / 2.0) <= x <= (
+            self.rz_x_ub - self._screw_width / 2.0)
+        return x_is_valid and y_is_valid
+
+    def _transition_move(self, state: State, x: float, y: float) -> State:
+        next_state = state.copy()
+        rx = state.get(self.robot, "pose_x")
+        ry = state.get(self.robot, "pose_y")
+        rwidth = state.get(self.robot, "width")
+        new_rx = rx + x
+        new_rx_min = new_rx - rwidth / 2.0
+        new_rx_max = new_rx + rwidth / 2.0
+        new_ry = ry + y
+
+        # If the new position is outside the boundaries, move till
+        # the robot collides with the boundary.
+        if new_rx_min < self.rz_x_lb:
+            new_rx = self.rz_x_lb + (rwidth / 2.0)
+        if new_rx_max > self.rz_x_ub:
+            new_rx = self.rz_x_ub - (rwidth / 2.0)
+        if new_ry < self.rz_y_lb:
+            new_ry = self.rz_y_lb
+        if new_ry > self.rz_y_ub:
+            new_ry = self.rz_y_ub
+        next_state.set(self.robot, "pose_x", new_rx)
+        next_state.set(self.robot, "pose_y", new_ry)
+
+        # Recompute dx and dy in case it's changed.
+        dx = new_rx - rx
+        dy = new_ry - ry
+
+        # Find out which (if any) screws are currently being held by
+        # the gripper and update their positions to follow the gripper.
+        all_screws = state.get_objects(self._screw_type)
+        for screw in all_screws:
+            if self._HoldingScrew_holds(state, [screw]):
+                screw_x = state.get(screw, "pose_x")
+                screw_y = state.get(screw, "pose_y")
+                next_state.set(screw, "pose_x", screw_x + dx)
+                next_state.set(screw, "pose_y", screw_y + dy)
+
+        return next_state
+
+    def _transition_magnetize(self, state: State) -> State:
+        next_state = state.copy()
+        ry = state.get(self.robot, "pose_y")
+        all_screws = state.get_objects(self._screw_type)
+        for screw in all_screws:
+            if not self._HoldingScrew_holds(state, [screw]):
+                # Check if the gripper is in the right position to
+                # pick up this screw, and if so, move the screw to
+                # the gripper.
+                if self._ScrewPickupable_holds(state, [self.robot, screw]):
+                    screw_height = state.get(screw, "height")
+                    next_state.set(screw, "pose_y", ry - (screw_height / 2.0))
```

Comment:
probably related to my rendering gripe, if `ry` is the center of the gripper, this calculation doesn't look quite right

Reply:
wait why? I feel like this is exactly right (`ry` is the center of the gripper, and `"pose_y"` is the center of the block, so we want to set `"pose_y"` to `ry - half_of_screw_height`)

Reply:
don't you need to take into account the height of the gripper?

Reply:
The gripper is a flat line and so doesn't have a height - I think during rendering it's given a height because of the line thickness

Reply:
ah, got it. in that case, you can just change the rendering. probably easier to make it a rectangle